stan/math/prim/mat/fun/value_of_rec.hpp

```diff
@@ -22,12 +22,33 @@ namespace stan {
     template <typename T, int R, int C>
     inline Eigen::Matrix<double, R, C>
     value_of_rec(const Eigen::Matrix<T, R, C>& M) {
-      using stan::math::value_of_rec;
-      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
-      for (int j = 0; j < M.cols(); j++)
-        for (int i = 0; i < M.rows(); i++)
-          Md(i, j) = value_of_rec(M(i, j));
-      return Md;
+      Eigen::Matrix<double, R, C>
+        result(M.rows(), M.cols());
+
+      double * data_r = result.data();
+      const T * data_m = M.data();
+
+      int S = M.size();
+      for (int i=0; i < S; i++)
+        data_r[i] = value_of_rec(data_m[i]);
+      return result;
+    }
```

Comment:
On 03/03/2016 20:35, Bob Carpenter wrote:
>
> In stan/math/prim/mat/fun/value_of_rec.hpp 
> <https://github.com/stan-dev/math/pull/250#discussion_r54967605>:
>
> > -      Eigen::Matrix<double, R, C> Md(M.rows(), M.cols());
> > -      for (int j = 0; j < M.cols(); j++)
> > -        for (int i = 0; i < M.rows(); i++)
> > -          Md(i, j) = value_of_rec(M(i, j));
> > -      return Md;
> > +      Eigen::Matrix<double, R, C>
> > +        result(M.rows(), M.cols());
> > +
> > +      double * data_r = result.data();
> > +      const T * data_m = M.data();
> > +
> > +      int S = M.size();
> > +      for (int i=0; i < S; i++)
> > +        data_r[i] = value_of_rec(data_m[i]);
> > +      return result;
> > +    }
>
> Well, maybe write out the time taken in more precision
>
Do you know how to do this with std::cout and std::time??

     cout << time(0) - now << endl;

(I don't think it'll be necessary thought, see below)

> or run more loops in the micro-benchmark.
>
If I try to double the number of elements in the matrix, I get 
std::bad_alloc.

But anyway, the 0's were coming up because the compiler optim realized 
that I wasn't using the output of the function at all, so it wasn't even 
running it!

So I changed;

     now = time(0);
     sum_kahan1t(test);
     cout << time(0) - now << endl;

to:

     now = time(0);
     cout << sum_kahan1t(test) << endl;
     cout << time(0) - now << endl;


And after running this a lot of times, I don't think there's any 
difference between them, and even if there is, it's probably negligible 
(iff -DEIGEN_NO_DEBUG and -O3 are used together).

$ g++ main.cpp -DEIGEN_NO_DEBUG `pkg-config eigen3 --cflags --libs` -O3
$ ./a.out
start
7697.42
3
7697.42
9
7697.42
4
7697.42
3

$ ./a.out
start
7697.42
3
7697.42
9
7697.42
3
7697.42
4


----------------------------------------------------

10 times run summed up:

     double te = 0;

     now = time(0);
     for (int i = 0; i < 10; i++)
       te += sum_kahan1t(test);
     cout << time(0) - now << endl;

     now = time(0);
     for (int i = 0; i < 10; i++)
       te += sum_kahan2t(test);
     cout << time(0) - now << endl;

     now = time(0);
     for (int i = 0; i < 10; i++)
       te += sum_kahan3t(test);
     cout << time(0) - now << endl;

     now = time(0);
     for (int i = 0; i < 10; i++)
       te += sum_kahan4t(test);
     cout << time(0) - now << endl;

     cout << te << endl;


$ g++ main.cpp -DEIGEN_NO_DEBUG `pkg-config eigen3 --cflags --libs` -O3
$ ./a.out
start
33
86
32
33
307897